### PR TITLE
Remove boost spirit from StringCutObjectParser

### PR DESCRIPTION
**Pull request by lgray**
This is a PR so that I can make sure this migration remains technical (through jenkins tests), to mark progress, and to serve as a warning for the coming change. 
Please do not merge this, it is not ready.

This PR a checkpoint and is the first step in removing boost::spirit (through ExpressionParser and various other things) from CMSSW.

StringCutObjectParser now does not use reco::parser, but instead reco::ExpressionEvaluator from @VinInn.
All string cuts passed to the parser are now required to be valid C++, I wrote a simple customize for the HLT to deal with this. Elsewhere I have changed the strings by hand.

Right now this switch results in very slow CMSSW run times, due to invoking the compiler.
No memory improvements seen yet since boost::spirit (and resulting calls to TCling) still hangs around in memory.

Also, this breaks igprof as it exists right now.

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by fwyzard** [comment]
what is the impact of this change on
  - execution time
  - start-up time
  - multi-threading
?

**Comment by Martin-Grunewald** [comment]
Please put the PR number in the HLT custom file, like it is done for the other customisations, otherwise we loose track!

**Comment by davidlt** [comment]
@fwyzard `CommonTools/Utils/src/ExpressionEvaluator.cc` there are few shell execution from inside C++ and also invocation of GCC driver. This is quite expensive operation, remember driver invokes few more tools. E.g., GCC JIT used a similar way (invoking _as_ and _ld_) and that was the biggest cost of JIT'ing.

Also ExpressionEvaluator involves PCH, but that part is compiler specific and e.g. fails on Clang (PCH (internal compiler dump) are not compatible).

Why not to use TCling for JIT'ing? Think, TFormula, which is also TCling based in ROOT 6.04. There are also examples how to use Clang directly for JIT'ing: https://github.com/llvm-mirror/clang/tree/master/examples/clang-interpreter

File: HLTrigger/Configuration/python/customizeHLTforMC.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
   process=customizeHLTforMC(process,_fastSim)
+  process=customizeHLTforExprEvaluator(process)
   return process
 
 def customizeHLTforFullSim(process,_fastSim=False):
   process=customizeHLTforMC(process,_fastSim)
+  process=customizeHLTforExprEvaluator(process)
   return process
 
 def customizeHLTforMC(process,_fastSim=False):
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTforMC.py]: Changes to customizeHLTforMC.py are indeed not needed. 
customizeHLTforCMSSW.py is run for both data and MC.

**Comment by lgray** [comment]
@davidlt @fwyzard Thanks for the comments. We were thinking to try to minimize use of clang::SEMA or other relatively memory-expensive things, since we're trying to kill a rather bulky parser. 

However, indeed, since we already use clang when reading the DataFormats perhaps we don't incur too much of a further penalty. I can give it a try, I think I will go for the direct clang interpreter, ROOT is always expensive.

On startup time:
Yes, this is indeed very very slow (as mentioned in the PR description), HLT startup is not very much affected 30s longer or so. RECO start up is 5x longer than normal. Clearly this is a non-starter for integration as it is, but the first step is showing you can do it! Comments so far have been very helpful.

Execution time:
Since this breaks igprof and incurs a huge amount of startup time I have not been able to measure the impact here in an accurate way. Execution time should improve, even if we are calling through a virtual function, since we no longer are using a boost::Grammar to parse a string into valid C++ which is then TCling'd.

Multithreading:
This PR makes StringCutObjectParser thread safe, as opposed to thread-friendly. You may use it in global modules now.

**Comment by lgray** [comment]
Looks like there are no changes against the baseline. @slava77 @cvuosalo could you comment here?

**Comment by davidlt** [comment]
@lgray could you elaborate how it breaks IgProf?

How often does CutParser expressions change? Does it make sense to create a cache (compile-time) and get already recompiled code if exist and not invoke compiler, assembler, linker? If those change rarely on production jobs you don't loose a lot on time.

BTW, this needs to be updated to work in Clang and ICC. That involves setting correct flags, PCH generation and detecting current compiler IIRC.

You can see it failing with Clang here: https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc493/CMSSW_7_6_CLANG_X_2015-09-15-2300/unitTestLogs/CommonTools/Utils

**Comment by lgray** [comment]
@davidlt The caching of expressions is already done. So that speed up in included (!!!) in the start up times we have right now. Clearly a faster baseline solution is needed.

It breaks igprof when it does the dlopen() after opening a child thread to do the compilation. This is again fixed if we move to a clang-based JIT.

**Comment by fwyzard** [comment]
> @davidlt @fwyzard Thanks for the comments. We were thinking to try to minimize use of clang::SEMA or other relatively memory-expensive things, since we're trying to kill a rather bulky parser. 

just to understand: the reason for the migration is to drop the dependency on boost::spirit ?

**Comment by lgray** [comment]
@davidlt just to make it clear, the real bottleneck comes from having to compile some 300 unique expressions, if DQM is involved. RECO + MiniAOD alone is faster, but still there GenParticlePruner is used in MC and that creates a new StringCutObjectParser for each line entered into it. Again after this first time it is compiled, everything is cached (which is why it doesn't take ages while running events).

**Comment by lgray** [comment]
@fwyzard Yes, this is the goal. This 'checkpoint' PR only achieves this for StringCutObjectParser, there are other things in CMSSW that use boost::spirit which must be addressed. (again this PR will not be merged in its present form, it is more to announce that progress is being made)

**Comment by lgray** [comment]
Also of note, there are a lot of places in the DQM code which use string cuts to implement full IDs. 
These use cases are likely better suited by VID (where possible), perhaps a migration campaign for that would be useful too.

**Comment by davidlange6** [comment]
> On Sep 17, 2015, at 11:01 AM, Andrea Bocci <notifications@github.com> wrote:
> 
> @davidlt @fwyzard Thanks for the comments. We were thinking to try to minimize use of clang::SEMA or other relatively memory-expensive things, since we're trying to kill a rather bulky parser.
> 
> just to understand: the reason for the migration is to drop the dependency on boost::spirit ?

the large memory footprint we get from using it (in root6 based releases).

**Comment by slava77** [comment]
indeed, there appears to be no changes visible in jenkins comparisons.

**Comment by Dr15Jones** [comment]
David is correct, the problem is not boost::spirit, the problem is the technology we use to associate a string name to a member function. The present technology we use is ROOT. This wasn't a problem in ROOT 5 since we already had the dictionaries generated and they didn't eat up too much memory. But in ROOT 6, the full dictionaries aren't generated until requested at run time and that takes a very large amount of memory and a considerable amount of time.
Now the present design of the cut parser does not require that we use ROOT to do the 'string to function' connection. It should be possible to pre-generate such a lookup (by hand or by automated script) and then have the parser first see if a pre-generated version exists and use it if found. This would completely avoid any run time overhead of calling a compiler and linker.

**Comment by lgray** [comment]
@Dr15Jones The only thing that this doesn't provide a nice workaround for are lazy-evaluated expressions.

**Comment by VinInn** [comment]
we shall get rid of lazy-evaluated expressions in production.
we shall collect all code-to-be-compiled in a single file, compile,load ONCE.
We shall forbid code-as-string at runtime past "initialization"

**Comment by lgray** [comment]
Do you have seven more?

**Comment by VinInn** [comment]
only the first 3 were relevant anyhow.
the rest is just derived...

**Comment by lgray** [comment]
@VinInn More seriously, I do think all of the lazy-evaluated statements that run in production at the moment can be made non-lazy. I'll have a look. The only place where the lazy evaluation is necessary (as far as I can tell) is fireworks.

**Comment by davidlt** [comment]
I do agree on @VinInn suggestions in particular for production jobs. Remember this would save 200+MB on RSS and maybe simply the code (?). I can dig up old IgProf reports showing CutParser dominating in RSS in ROOT5 vs ROOT6 comparison.

**Comment by gpetruc** [comment]
If the backend is the problem, can we try fix or replace that without breaking the grammar, which otherwise probably breaks also a very large amount of analysis code? 

Essentially everything that is in production offline could be precompiled when building the release since the string defining the expression or cut is available in the release (online maybe not because code and config are in two different places)

**Comment by VinInn** [comment]
fireworks is a completely different story.
It is interactive, its latency affects only one user who is clicking and typing (that's slow anyhow) 
It can be fully written in python or similar...

**Comment by fwyzard** [comment]
> If the backend is the problem, can we try fix or replace that without breaking the grammar, which otherwise probably breaks also a very large amount of analysis code?

+1

> Essentially everything that is in production offline could be precompiled when building the release since the string defining the expression or cut is available in the release 

> (online maybe not because code and config are in two different places)

luckily, or we would need a new release every few hours...

**Comment by lgray** [comment]
Hmm, this is likely possible. I will think about it.

**Comment by gpetruc** [comment]
If you change the backend at the level where it resolves individual atoms and not entire expressions, then you can use a library of functions again.
I.e. for a call "daughter.pt" you'd only need the typeid of what's returned by "daughter" and lookup what "pt" does for that.

**Comment by lgray** [comment]
OK - Thanks for the suggestions, closing this PR. It will return later after trying these suggestions out.

**Comment by lgray** [comment]
OK - found discussion in the T0 HN regarding things related to this. Will follow up from there.

**Comment by slava77** [comment]
On 9/18/15 3:05 AM, Giovanni Petrucciani wrote:
> If the backend is the problem, can we try fix or replace that without
> breaking the grammar, which otherwise probably breaks also a very large
> amount of analysis code?
> 
> Essentially everything that is in production offline could be
> precompiled when building the release since the string defining the
> expression or cut is available in the release (online maybe not because
> code and config are in two different places)

Some formulas are defined in conditions payloads.
So, it's not really possibly to pre-compile these, they are discovered
only runtime.

**Comment by VinInn** [comment]
> On 18 Sep, 2015, at 3:10 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 9/18/15 3:05 AM, Giovanni Petrucciani wrote:
> > If the backend is the problem, can we try fix or replace that without
> > breaking the grammar, which otherwise probably breaks also a very large
> > amount of analysis code?
> > 
> > Essentially everything that is in production offline could be
> > precompiled when building the release since the string defining the
> > expression or cut is available in the release (online maybe not because
> > code and config are in two different places)
> 
> Some formulas are defined in conditions payloads.
> So, it's not really possibly to pre-compile these, they are discovered
> only runtime.
as commented above this is even easier to fix…
It is just a bit of (good practice) effort to code c++ in code not in std::string
v.

**Comment by Dr15Jones** [comment]
I completely agree with Giovanni and is what I was trying to say in my original comment. We can just replace the lookup for 'daughter.pt' to look at our own list instead of calling into ROOT. We already know that the expression evaluation in the cut parser is sufficiently efficient to not cause timing problems when run single-threaded. The underlying problem we are presently seeing is the use of ROOT to resolve names to function.

**Comment by lgray** [comment]
@Dr15Jones In your thread on the HN it seems you are already working on something towards this goal? Anyway I can help?

**Comment by Dr15Jones** [comment]
@lgray I'm working on a replacement for TFormula (pretty far along after only one evening). I haven't spent any time on the cut parser. I did a few weeks back look to see if replacing ROOT was at all feasible and it did appear so. If you want to look at the cut parser code that looks functions that would be awesome.

**Comment by lgray** [comment]
@Dr15Jones I will take a look and see where I get. Fixing EGM things first though.

**Pull request closed without merging**